core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
+      val data = ex(j).asInstanceOf[Object]
+//      println(s"@$i/$j = $data")
+      data
+    }
+
+    def extractRow: Option[Array[AnyRef]] = {
+      val rowOpt = opt {
+        val objects: IndexedSeq[Object] = columnTypeNames.indices map cell
+        val row = objects toArray
+
+//        println(s"${converterCtx.rowIdx} -> ${res mkString ":"}")
+        row
+      }
+      converterCtx.increaseRowIdx()
+      rowOpt
+    }
+
+    private var hd: Option[A] = None
+    private var total = 0
 
-      def next(): A = {
-        val data = new Array[Option[Any]](numCols)
-        // FIXME: this is not perfect since ncols does not need to match number of names
-        (0 until numCols).foreach{ idx =>
-          val value = if (converterCtx.isNA(idx)) None
-          else types(idx) match {
-            case q if q == classOf[Integer]           => Some(converterCtx.getInt(idx))
-            case q if q == classOf[java.lang.Long]    => Some(converterCtx.getLong(idx))
-            case q if q == classOf[java.lang.Double]  => Some(converterCtx.getDouble(idx))
-            case q if q == classOf[java.lang.Float]   => Some(converterCtx.getFloat(idx))
-            case q if q == classOf[java.lang.Boolean] => Some(converterCtx.getBoolean(idx))
-            case q if q == classOf[String] => Option(converterCtx.getString(idx))
-            case _ => None
+    override def hasNext = {
+      while (hd.isEmpty && super.hasNext) {
+        hd = readOne()
+        total += 1
+      }
+      hd.isDefined
+    }
+
+    def next(): A = {
+      if (hasNext) {
+        val a = hd.get
+        hd = None
+        a
+      } else {
+        throw new NoSuchElementException(s"No more elements in this iterator: found $total  out of ${converterCtx.numRows}")
+            }
+      }
+
+      private def readOne(): Option[A] = {
+            val dataOpt = extractRow
+
+            val res: Seq[A] = for {
+            builder <- builders
+            data <- dataOpt
+            instance <- builder(data)
+            } yield instance
+
+            res.toList match {
+              case Nil => None
+              case unique :: Nil => Option(unique)
+              case one :: two :: more =>
+                throw new scala.IllegalArgumentException(
+                                         s"found more than une $jc constructor for given args - can't choose")
           }
-          data(idx) = value
         }
 
-        converterCtx.increaseRowIdx()
-        // Create instance for the extracted row
-        ccr.newInstance(data:_*).asInstanceOf[A]
-      }
+    override lazy val converterCtx: ReadConverterContext =
+      ConverterUtils.getReadConverterContext(keyName,
+        partIndex)
+  }
+
+
+  lazy val constructors: Seq[Constructor[_]] = {
+
+    val cs = jc.getConstructors
+    val found = cs.collect {
+      case c if c.getParameterTypes.length == colNamesInResult.length => c
     }
 
-    ConverterUtils.getIterator[A](isExternalBackend, iterator)
```

Comment:
Done; thanks!

Thanks,
-Vlad

On Fri, Aug 26, 2016 at 12:42 AM, Jakub Háva <notifications@github.com>
wrote:

> In core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala
> <https://github.com/h2oai/sparkling-water/pull/58#discussion_r76376766>:
>
> >      }
> >
> > -    ConverterUtils.getIterator[A](isExternalBackend, iterator)
>
> this line ConverterUtils.getIterator[A](isExternalBackend, iterator) is
> in the master, at the end of H2ORDD branch.
>
> You can either keep this method call there, or I will have to add this
> call there as part of integration with SW-178.
> For Internal backend ( this one ) it just forwards the iterator, for the
> external it has to do some buffering.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2oai/sparkling-water/pull/58/files/dda8bacb29caf9d9ad724922e0bdf29e831b7ecb#r76376766>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOuK3g_pdKUauzx-OHObiDydMdE2NI3ks5qjpjMgaJpZM4Jsgi_>
> .
>